File: clients/go/pool.go
package pulsedb

import (
	"crypto/tls"
	"fmt"
	"sync"
)

// Pool is a thread-safe connection pool for PulseDB.
//
// Usage:
//
//	pool, err := pulsedb.NewPool(pulsedb.PoolConfig{
//	    Addr:     "127.0.0.1:7878",
//	    MinSize:  2,
//	    MaxSize:  10,
//	    Username: os.Getenv("PULSEDB_USER"),
//	    Password: os.Getenv("PULSEDB_PASSWORD"),
//	})
//	if err != nil { log.Fatal(err) }
//	defer pool.Close()
//
//	conn, err := pool.Acquire()
//	if err != nil { ... }
//	defer pool.Release(conn)
//	conn.Query("GET users")
type Pool struct {
	cfg  PoolConfig
	idle chan *Client
	mu   sync.Mutex
	open int
}

// PoolConfig holds the settings for a connection pool.
type PoolConfig struct {
	// Addr is the server address, e.g. "127.0.0.1:7878".
	Addr string
	// MinSize is the number of connections opened eagerly.
	MinSize int
	// MaxSize is the hard cap on total connections.
	MaxSize int
	// Username and Password for authentication (leave empty for --no-auth servers).
	Username string
	Password string
	// TLSConfig enables TLS when non-nil.
	// Use &tls.Config{InsecureSkipVerify: true} for self-signed certs in dev.
	TLSConfig *tls.Config
}

// NewPool creates a pool and opens MinSize connections eagerly.
func NewPool(cfg PoolConfig) (*Pool, error) {
	if cfg.MaxSize <= 0 {
		cfg.MaxSize = 10
	}
	if cfg.MinSize <= 0 {
		cfg.MinSize = 1
	}
	if cfg.MinSize > cfg.MaxSize {
		cfg.MinSize = cfg.MaxSize
	}

	p := &Pool{
		cfg:  cfg,
		idle: make(chan *Client, cfg.MaxSize),
	}

	for i := 0; i < cfg.MinSize; i++ {
		c, err := p.dial()
		if err != nil {
			p.Close()
			return nil, fmt.Errorf("pulsedb pool: initial connect: %w", err)
		}
		p.idle <- c
		p.mu.Lock()
		p.open++
		p.mu.Unlock()
	}
	return p, nil
}

// Acquire checks out a connection. Call Release when done.
func (p *Pool) Acquire() (*Client, error) {
	select {
	case c := <-p.idle:
		if p.ping(c) {
			return c, nil
		}
		// Stale — replace it
		c.Close()
		p.mu.Lock()
		p.open--
		p.mu.Unlock()
	default:
	}

	// No idle connection — open a new one if under the cap
	p.mu.Lock()
	if p.open < p.cfg.MaxSize {
		p.open++
		p.mu.Unlock()
		c, err := p.dial()
		if err != nil {
			p.mu.Lock()
			p.open--
			p.mu.Unlock()
			return nil, fmt.Errorf("pulsedb pool: connect: %w", err)
		}
		return c, nil
	}
	p.mu.Unlock()

	// At cap — wait for a connection to be released
	c := <-p.idle
	if !p.ping(c) {
		c.Close()
		p.mu.Lock()
		p.open--
		p.mu.Unlock()
		return p.Acquire() // retry once
	}
	return c, nil
}

// Release returns a connection to the pool.
func (p *Pool) Release(c *Client) {
	if c == nil {
		return
	}
	select {
	case p.idle <- c:
	default:
		// Pool is full — close the excess connection
		c.Close()
		p.mu.Lock()
		p.open--
		p.mu.Unlock()
	}
}

// Close shuts down every connection in the pool.
func (p *Pool) Close() {
	close(p.idle)
	for c := range p.idle {
		c.Close()
	}
}

// Stats returns (idle, total) connection counts.
func (p *Pool) Stats() (idle, total int) {
	p.mu.Lock()
	total = p.open
	p.mu.Unlock()
	idle = len(p.idle)
	return
}

func (p *Pool) dial() (*Client, error) {
	var (
		c   *Client
		err error
	)
	if p.cfg.TLSConfig != nil {
		c, err = ConnectTLS(p.cfg.Addr, p.cfg.TLSConfig)
	} else {
		c, err = Connect(p.cfg.Addr)
	}
	if err != nil {
		return nil, err
	}
	if p.cfg.Username != "" && p.cfg.Password != "" {
		if authErr := c.Auth(p.cfg.Username, p.cfg.Password); authErr != nil {
			c.Close()
			return nil, authErr
		}
	}
	return c, nil
}

func (p *Pool) ping(c *Client) bool {
	_, err := c.Query("SHOW TABLES")
	return err == nil
}

File: clients/javascript/pool.js
'use strict';

/**
 * PulseDB connection pool for Node.js.
 *
 * Manages a pool of reusable PulseDB connections — essential for servers
 * handling many concurrent requests.
 *
 * @example
 * ```js
 * const { ConnectionPool } = require('./pool');
 *
 * const pool = new ConnectionPool({
 *   host: '127.0.0.1', port: 7878,
 *   minSize: 2, maxSize: 10,
 *   username: process.env.PULSEDB_USER,
 *   password: process.env.PULSEDB_PASSWORD,
 * });
 *
 * // Acquire → use → auto-release
 * const result = await pool.withConnection(async (db) => {
 *   return db.query('GET users LIMIT 10');
 * });
 *
 * // Or manually:
 * const db = await pool.acquire();
 * try {
 *   await db.query('PUT users { id: 1, name: "Alice" }');
 * } finally {
 *   pool.release(db);
 * }
 *
 * await pool.close();
 * ```
 */

const { PulseDB } = require('./index');

class ConnectionPool {
  /**
   * @param {object} opts
   * @param {string}  [opts.host='127.0.0.1']
   * @param {number}  [opts.port=7878]
   * @param {number}  [opts.minSize=1]      Connections opened at construction
   * @param {number}  [opts.maxSize=10]     Hard cap on total connections
   * @param {string}  [opts.username]
   * @param {string}  [opts.password]
   * @param {boolean} [opts.tls=false]
   * @param {boolean} [opts.tlsNoVerify=false]
   * @param {number}  [opts.acquireTimeout=30000]  ms to wait for a free slot
   */
  constructor(opts = {}) {
    this._host           = opts.host           || '127.0.0.1';
    this._port           = opts.port           || 7878;
    this._minSize        = opts.minSize        || 1;
    this._maxSize        = opts.maxSize        || 10;
    this._username       = opts.username       || null;
    this._password       = opts.password       || null;
    this._tls            = opts.tls            || false;
    this._tlsNoVerify    = opts.tlsNoVerify    || false;
    this._acquireTimeout = opts.acquireTimeout || 30_000;

    this._idle    = [];   // available connections
    this._open    = 0;    // total connections (idle + in-use)
    this._waiting = [];   // {resolve, reject, timer} queue
    this._closed  = false;
    this._ready   = this._init();
  }

  // ── Initialisation ────────────────────────────────────────────────────

  async _init() {
    for (let i = 0; i < this._minSize; i++) {
      const c = await this._dial();
      this._idle.push(c);
      this._open++;
    }
  }

  /** Wait for the pool to be ready (minSize connections open). */
  async ready() { return this._ready; }

  // ── Public API ────────────────────────────────────────────────────────

  /**
   * Acquire a connection from the pool.
   * Resolves immediately if one is idle, or waits up to `acquireTimeout` ms.
   * @returns {Promise<PulseDB>}
   */
  acquire() {
    if (this._closed) return Promise.reject(new Error('pool is closed'));

    // Fast path: idle connection available
    if (this._idle.length > 0) {
      return Promise.resolve(this._idle.pop());
    }

    // Can open a new connection
    if (this._open < this._maxSize) {
      this._open++;
      return this._dial().catch(err => {
        this._open--;
        throw err;
      });
    }

    // At cap — queue the request
    return new Promise((resolve, reject) => {
      const timer = setTimeout(() => {
        const idx = this._waiting.findIndex(w => w.resolve === resolve);
        if (idx !== -1) this._waiting.splice(idx, 1);
        reject(new Error(
          `connection pool exhausted (max=${this._maxSize}) — ` +
          `increase maxSize or acquireTimeout`
        ));
      }, this._acquireTimeout);

      this._waiting.push({ resolve, reject, timer });
    });
  }

  /**
   * Return a connection to the pool.
   * @param {PulseDB} conn
   */
  release(conn) {
    if (!conn) return;

    // Serve any waiting acquirer first
    if (this._waiting.length > 0) {
      const { resolve, timer } = this._waiting.shift();
      clearTimeout(timer);
      resolve(conn);
      return;
    }

    // Verify the connection is still alive
    conn.query('SHOW TABLES').then(() => {
      this._idle.push(conn);
    }).catch(() => {
      conn.close();
      this._open--;
    });
  }

  /**
   * Run `fn(conn)` with an auto-released connection.
   * @template T
   * @param {function(PulseDB): Promise<T>} fn
   * @returns {Promise<T>}
   */
  async withConnection(fn) {
    const conn = await this.acquire();
    try {
      return await fn(conn);
    } finally {
      this.release(conn);
    }
  }

  /**
   * Close all connections and shut down the pool.
   */
  async close() {
    this._closed = true;
    for (const { reject, timer } of this._waiting) {
      clearTimeout(timer);
      reject(new Error('pool closed'));
    }
    this._waiting = [];
    for (const conn of this._idle) { conn.close(); }
    this._idle = [];
    this._open = 0;
  }

  /** Current pool statistics. */
  get stats() {
    return {
      idle:   this._idle.length,
      total:  this._open,
      waiting: this._waiting.length,
    };
  }

  // ── Internal ──────────────────────────────────────────────────────────

  async _dial() {
    const conn = await PulseDB.connect({
      host: this._host, port: this._port,
      tls: this._tls, tlsNoVerify: this._tlsNoVerify,
    });
    if (this._username && this._password) {
      await conn.auth(this._username, this._password);
    }
    return conn;
  }
}

module.exports = { ConnectionPool };
